pkg/api/composition.go

```diff
@@ -273,294 +310,195 @@ type Dependency struct {
+	for _, runId := range runIds {
+		run, err := c.getRun(runId)
 
-	// Trickle global build defaults to groups, if any.
-	if def := c.Global.Build; def != nil {
-		for _, grp := range c.Groups {
-			grp.Build.Dependencies = grp.Build.Dependencies.ApplyDefaults(def.Dependencies)
-			if len(grp.Build.Selectors) == 0 {
-				grp.Build.Selectors = def.Selectors
-			}
+		if err != nil {
+			return nil, fmt.Errorf("invalid run id %s: %w", runId, err)
 		}
-	}
 
-	// Trickle global build config to groups, if any.
-	if len(c.Global.BuildConfig) > 0 {
-		for _, grp := range c.Groups {
-			if grp.BuildConfig == nil {
-				grp.BuildConfig = make(map[string]interface{})
-			}
-
-			for k, v := range c.Global.BuildConfig {
-				// Note: we only merge root values.
-				if _, ok := grp.BuildConfig[k]; !ok {
-					grp.BuildConfig[k] = v
-				}
-			}
+		for _, group := range run.Groups {
+			requiredGroupsIds[group.EffectiveGroupId()] = true
 		}
+
+		runs = append(runs, run)
 	}
 
-	// Trickle builder configuration
-	for _, grp := range c.Groups {
-		if grp.Builder == "" {
-			grp.Builder = c.Global.Builder
-		}
+	// Gather the groups that we listed in requiredGroupsIdx.
+	groups := make([]*Group, 0, len(requiredGroupsIds))
+	for groupId := range requiredGroupsIds {
+		group, err := c.GetGroup(groupId)
 
-		if !manifest.HasBuilder(grp.Builder) {
-			return nil, fmt.Errorf("plan does not support builder '%s'; supported: %v", c.Global.Builder, manifest.SupportedBuilders())
+		if err != nil {
+			return nil, fmt.Errorf("invalid group id %s: %w", groupId, err)
 		}
+
+		groups = append(groups, group)
 	}
 
+	c.Groups = groups
+	c.Runs = runs
+
 	return &c, nil
 }
 
-func (c *Composition) ListBuilders() []string {
-	builders := make(map[string]bool)
+func (c Composition) getRun(runId string) (*Run, error) {
+	for _, x := range c.Runs {
+		if x.ID == runId {
+			return x, nil
+		}
+	}
+	return nil, fmt.Errorf("unknown run id %s", runId)
+}
 
-	for _, grp := range c.Groups {
-		if grp.Builder == "" {
-			builders[c.Global.Builder] = true
-		} else {
-			builders[grp.Builder] = true
+func (c Composition) GetGroup(groupId string) (*Group, error) {
+	for _, x := range c.Groups {
+		if x.ID == groupId {
+			return x, nil
 		}
 	}
+	return nil, fmt.Errorf("unknown group id %s", groupId)
+}
 
-	result := make([]string, 0, len(builders))
-	for k := range builders {
-		result = append(result, k)
+func (c Composition) ListRunIds() []string {
+	ids := make([]string, 0, len(c.Runs))
+	for _, x := range c.Runs {
+		ids = append(ids, x.ID)
 	}
+	sort.Strings(ids)
+	return ids
+}
 
-	sort.Strings(result)
+func (c Composition) ListGroupsId() []string {
+	ids := make([]string, 0, len(c.Groups))
+	for _, x := range c.Groups {
+		ids = append(ids, x.ID)
+	}
+	sort.Strings(ids)
+	return ids
+}
 
-	return result
+// CalculatedInstanceCount returns the actual number of instances in this group.
+//
+// Validate MUST be called for this field to be available.
```

Comment:
Agreed,
but there's no "good" check for this other than rethink the whole thing into a getter rather than precompute it here.
(checking that the value is != 0 would hide bugs like "I changed the composition without recomputing indices")

I would like to simplify and refactor the composition format as a whole,
This PR is "only" adding a comment that makes this explicit, the issue existed before.